services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/LayeredKeyValueStorage.java

```diff
@@ -101,25 +102,72 @@ public Optional<byte[]> get(final SegmentIdentifier segmentId, final byte[] key)
-    Optional<NearestKeyValue> ourNearest = super.getNearestTo(segmentIdentifier, key);
-    Optional<NearestKeyValue> parentNearest = parent.getNearestTo(segmentIdentifier, key);
+    return getNearest(
+        key,
+        k -> super.getNearestBefore(segmentIdentifier, k),
+        k -> parent.getNearestBefore(segmentIdentifier, k),
+        false);
+  }
+
+  @Override
+  public Optional<NearestKeyValue> getNearestAfter(
+      final SegmentIdentifier segmentIdentifier, final Bytes key) throws StorageException {
+    return getNearest(
+        key,
+        k -> super.getNearestAfter(segmentIdentifier, k),
+        k -> parent.getNearestAfter(segmentIdentifier, k),
+        true);
+  }
+
+  private Optional<NearestKeyValue> getNearest(
+      final Bytes key,
+      final Function<Bytes, Optional<NearestKeyValue>> ourNearestFunction,
+      final Function<Bytes, Optional<NearestKeyValue>> parentNearestFunction,
+      final boolean isAfter)
+      throws StorageException {
+
+    Optional<NearestKeyValue> ourNearest = ourNearestFunction.apply(key);
+    Optional<NearestKeyValue> parentNearest = parentNearestFunction.apply(key);
 
     if (ourNearest.isPresent() && parentNearest.isPresent()) {
-      // Both are present, return the one closer to the key
-      int ourDistance = ourNearest.get().key().commonPrefixLength(key);
-      int parentDistance = parentNearest.get().key().commonPrefixLength(key);
-      return (ourDistance <= parentDistance) ? ourNearest : parentNearest;
+      // Both are present, return the one closer to the key or based on common prefix length
+      return compareNearest(ourNearest, parentNearest, key, isAfter);
     } else if (ourNearest.isPresent()) {
       // Only ourNearest is present
       return ourNearest;
     } else {
-      // return parentNearest, which may be an empty Optional
+      // Return parentNearest, which may be an empty Optional
       return parentNearest;
     }
   }
 
+  private Optional<NearestKeyValue> compareNearest(
+      final Optional<NearestKeyValue> ourNearest,
+      final Optional<NearestKeyValue> parentNearest,
+      final Bytes key,
+      final boolean isAfter) {
+
+    final int ourDistance = ourNearest.get().key().compareTo(key);
+    final int parentDistance = parentNearest.get().key().compareTo(key);
+    if (ourDistance == 0) {
+      return ourNearest;
+    } else if (parentDistance == 0) {
+      return parentNearest;
+    } else {
+      final int ourCommonPrefixLength = ourNearest.get().key().commonPrefixLength(key);
+      final int parentCommonPrefixLength = parentNearest.get().key().commonPrefixLength(key);
+      if (ourCommonPrefixLength != parentCommonPrefixLength) {
+        return ourCommonPrefixLength > parentCommonPrefixLength ? ourNearest : parentNearest;
+      } else {
+        return (isAfter ^ (ourNearest.get().key().compareTo(parentNearest.get().key()) > 0))
+            ? ourNearest
+            : parentNearest;
```

Comment:
this is very succinct 🥇, but I had to think hard about this.  It might be worth a comment to explain to our future selves

Reply:
Is it just this line that seems complicated or the method?

Reply:
I simplified this code and added explanation , feel free to tell me if it's better